packages/react-server/src/ReactFizzServer.js

```diff
@@ -97,119 +212,580 @@ function renderNode(request: Request, node: ReactNodeList): void {
+  const segment = work.blockedSegment;
+  const boundary = work.blockedBoundary;
   try {
-    let node = segment.node;
+    let node = work.node;
     while (
       typeof node === 'object' &&
       node !== null &&
       (node: any).$$typeof === REACT_ELEMENT_TYPE &&
       typeof node.type === 'function'
     ) {
-      // Attempt to render the server component.
-      // Doing this here lets us reuse this same segment if the next component
+      // Doing this here lets us reuse this same Segment if the next component
       // also suspends.
       const element: React$Element<any> = (node: any);
-      segment.node = node;
+      work.node = node;
       // TODO: Classes and legacy context etc.
       node = element.type(element.props);
     }
 
-    renderNode(request, node);
+    renderNode(request, boundary, segment, node);
+
+    completeWork(request, boundary, segment);
   } catch (x) {
     if (typeof x === 'object' && x !== null && typeof x.then === 'function') {
       // Something suspended again, let's pick it back up later.
-      const ping = segment.ping;
+      const ping = work.ping;
       x.then(ping, ping);
-      return;
     } else {
-      // This errored, we need to serialize this error to the
-      reportError(request, x);
-      // TODO: Emit client render signal
+      errorWork(request, boundary, segment, x);
     }
   }
 }
 
 function performWork(request: Request): void {
+  if (request.status === CLOSED) {
+    return;
+  }
   const prevDispatcher = ReactCurrentDispatcher.current;
   ReactCurrentDispatcher.current = Dispatcher;
 
-  const pingedSegments = request.pingedSegments;
-  request.pingedSegments = [];
-  for (let i = 0; i < pingedSegments.length; i++) {
-    const segment = pingedSegments[i];
-    retrySegment(request, segment);
+  try {
+    const pingedWork = request.pingedWork;
+    let i;
+    for (i = 0; i < pingedWork.length; i++) {
+      const work = pingedWork[i];
+      retryWork(request, work);
+    }
+    pingedWork.splice(0, i);
+    if (request.status === FLOWING) {
+      flushCompletedQueues(request);
+    }
+  } catch (error) {
+    fatalError(request, error);
+  } finally {
+    ReactCurrentDispatcher.current = prevDispatcher;
   }
-  if (request.flowing) {
-    flushCompletedChunks(request);
+}
+
+function flushSubtree(
+  request: Request,
+  destination: Destination,
+  segment: Segment,
+): boolean {
+  segment.parentFlushed = true;
+  switch (segment.status) {
+    case PENDING: {
+      // We're emitting a placeholder for this segment to be filled in later.
+      // Therefore we'll need to assign it an ID - to refer to it by.
+      segment.id = request.nextSegmentId++;
+      // TODO: Move this to host config and format a placeholder node.
+      const instruction = new Uint8Array(0);
+      return writeChunk(destination, instruction);
+    }
+    case COMPLETED: {
+      segment.status = FLUSHED;
+      let r = true;
```

Comment:
If it returns false, it indicates that the underlying buffer was full by the time we finished writing and we shouldn't write anything more. We'll still finish buffering up the remainder of this tree so that we have a complete tree but after that we should stop.

Reply:
Would be nice to clarify in the naming